### PR TITLE
Fix file not being deleted on rename + other `sync/end` bugs

### PR DESCRIPTION
**Pull request by james-wallis**
Fixes https://github.com/eclipse/codewind/issues/1289 (File not being deleted on rename)
### Summary
* Fixed issue with file not being deleted after a rename
* Fixed unknown issue about file not being deleted when no file is changed / a new file is uploaded
  * If a file was not uploaded with sync then the `upload/end` API would return 404 without checking whether the modified list had any files inside thus skipped deleting files.
* Fixed unknown issue about file not being deleted when there is a space in the filename - changed sharedFunctions.forceRemove 

### Testing
* Added unit tests for functions introduced into `remoteBind.route.js` to sanity check them.
* Added unit tests for an existing function.
* Added unit tests for `sharedFunctions.forceRemove()`

### CODE REVIEW

File: src/pfe/portal/routes/projects/remoteBind.route.js [hunk 1]
```diff
@@ -212,24 +212,23 @@ router.post('/api/v1/projects/:id/upload/end', async (req, res) => {
...
       } else {
-
-        const currentFileList = await listFiles(pathToTempProj, '');
-
-        const filesToDeleteSet = new Set(currentFileList);
-        keepFileList.forEach((f) => filesToDeleteSet.delete(f));
-        const filesToDelete = Array.from(filesToDeleteSet);
+        const pathToProj = path.join(global.codewind.CODEWIND_WORKSPACE, project.name);
+        const currentFileList = await listFilesInDirectory(pathToProj);
+        // don't use pathToTempProj use the actual path ..... obv
```
Comment (hhellyer) [inline src/pfe/portal/routes/projects/remoteBind.route.js]: The meaning of obv is obvious obviously...
(Tidy up the comment?)

File: src/pfe/portal/routes/projects/remoteBind.route.js [hunk 2]
```diff
@@ -334,6 +342,10 @@ async function listFiles(absolutePath, relativePath) {
   return fileList;
 }
 
+function listFilesInDirectory(absolutePathToDirectory) {
```
Comment (hhellyer) [inline src/pfe/portal/routes/projects/remoteBind.route.js]: You could have achieved the same effect by setting a default value in listFiles:
```
async function listFiles(absolutePath, relativePath = '') {
```

File: src/pfe/portal/routes/projects/remoteBind.route.js [hunk 3]
```diff
@@ -212,24 +212,23 @@ router.post('/api/v1/projects/:id/upload/end', async (req, res) => {
...
+        log.info('Temporary project directory doesn\'t exist and modified list is empty, not syncing any files');
+        res.status(404).send('No files have been synced');
       } else {
-
-        const currentFileList = await listFiles(pathToTempProj, '');
-
-        const filesToDeleteSet = new Set(currentFileList);
-        keepFileList.forEach((f) => filesToDeleteSet.delete(f));
-        const filesToDelete = Array.from(filesToDeleteSet);
+        const pathToProj = path.join(global.codewind.CODEWIND_WORKSPACE, project.name);
```
Comment (hhellyer) [inline src/pfe/portal/routes/projects/remoteBind.route.js]: Should you use project.projectPath() here? I think there may be cases the project directory doesn't match the name. They are certainly different fields in the project object.

File: src/pfe/portal/routes/projects/remoteBind.route.js [hunk 4]
```diff
@@ -212,24 +212,23 @@ router.post('/api/v1/projects/:id/upload/end', async (req, res) => {
     const project = user.projectList.retrieveProject(projectID);
     if (project) {
       const pathToTempProj = path.join(global.codewind.CODEWIND_WORKSPACE, global.codewind.CODEWIND_TEMP_WORKSPACE, project.name);
+      const tempProjectExists = await fs.pathExists(pathToTempProj);
       // eslint-disable-next-line no-sync
```
Comment (rwalle61) [inline src/pfe/portal/routes/projects/remoteBind.route.js]: don't need this anymore

File: src/pfe/portal/routes/projects/remoteBind.route.js [hunk 5]
```diff
@@ -320,7 +329,6 @@ async function listFiles(absolutePath, relativePath) {
     const nextRelativePath = path.join(relativePath, f);
```
Comment (rwalle61) [inline src/pfe/portal/routes/projects/remoteBind.route.js]: perhaps out of scope but fyi there are open source node modules that do this function (better than us) e.g. https://www.npmjs.com/package/recursive-readdir-async

File: test/src/unit/routes/projects/remoteBind.route.test.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+function createDirectoriesFromArray(directory, dirArray) {
+    const promiseArray = dirArray.map(dir => {
+        return fs.ensureDir(path.join(directory, dir));
+    });
+    return Promise.all(promiseArray);
+}
+
+function assertFilesExist(directory, fileArray) {
+    for (let i = 0; i < fileArray.length; i++) {
```
Comment (rwalle61) [inline test/src/unit/routes/projects/remoteBind.route.test.js]:
```suggestion
    for (const relativePath of fileArray) {
```

**Pull request merged**